### PR TITLE
chore: add feature flag cache processor

### PR DESCRIPTION
**Pull request by cre8ivejp**
Part of https://github.com/bucketeer-io/bucketeer/issues/921

--- 

This pull request introduces changes to the feature flag caching and processing system in the `pkg/bucketeer/cache` package. The most significant changes include the addition of delete methods in the `FeaturesCache` interface, the creation of a new `FeatureFlagProcessor` interface and its implementation, and the corresponding tests for these new features.

Changes to the `FeaturesCache` interface:

* [`pkg/bucketeer/cache/features.go`](diffhunk://#diff-aec8f379bb114799fcdf2654ee61b1a92816780c082942a1a162805d0aab6757R37-R42): Added `Delete` and `DeleteAll` methods to the `FeaturesCache` interface. These methods allow for the removal of individual feature flags and all feature flags, respectively.
* [`pkg/bucketeer/cache/features.go`](diffhunk://#diff-aec8f379bb114799fcdf2654ee61b1a92816780c082942a1a162805d0aab6757R82-R97): Implemented the `Delete` and `DeleteAll` methods in the `featuresCache` struct.
* [`pkg/bucketeer/cache/mock/features.go`](diffhunk://#diff-92c78867694aca355ee3b1104dd76e92bc52ea873da7b22d5668195df1f312b2R37-R62): Added mock methods for `Delete` and `DeleteAll` in the `MockFeaturesCache` struct.

Creation of the `FeatureFlagProcessor`:

* [`pkg/bucketeer/cache/processor/feature_flag_processor.go`](diffhunk://#diff-39d4680964b0405085215e7c71c9dd8a26656177eace993d30e5f3e495802796R1-R231): Created a new `FeatureFlagProcessor` interface and its implementation. The processor is responsible for polling the latest feature flags from the server and updating them in the local cache.

Changes to tests:

* [`pkg/bucketeer/cache/features_test.go`](diffhunk://#diff-203679a19a621829c8e987eacd9a8c4f5358354d26e8d55865dbe7ad9301bf37R133-R171): Added a test for the `DeleteAll` method in the `FeaturesCache` interface.
* [`pkg/bucketeer/cache/processor/feature_flag_processor_test.go`](diffhunk://#diff-b2a32d9735aa23a42d7e2267ca6c7d6de5c26bb4f0a79767c251722089c63c03R1-R383): Created new tests for the `FeatureFlagProcessor`.

### CODE REVIEW

File: pkg/bucketeer/cache/processor/feature_flag_processor.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+func (p *processor) runProcessLoop() {
+	defer func() {
+		p.loggers.Debug("bucketeer/cache: runProcessLoop done")
+	}()
+	ticker := time.NewTicker(p.pollingInterval)
+	defer ticker.Stop()
+	// Update the cache once when starts polling
+	if err := p.updateCache(); err != nil {
```
Comment (cre8ivejp) [inline pkg/bucketeer/cache/processor/feature_flag_processor.go:90]: Here, we call once so it updates the cache right after the `Run` is called.

File: pkg/bucketeer/cache/processor/feature_flag_processor.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+		requestedAt,
+	)
+	// Get the latest cache from the server
+	resp, size, err := p.apiClient.GetFeatureFlags(req)
+	if err != nil {
+		return err
+	}
+	p.loggers.Debugf("bucketeer/cache: GetFeatureFlags response: %v, size: %d", resp, size)
+	// Delete all the local cache and save the new one
+	if resp.ForceUpdate {
```
Comment (cre8ivejp) [inline pkg/bucketeer/cache/processor/feature_flag_processor.go:133]: When the `forceUpdate` is true, it must delete all the local cache and save the new one from the response.

File: pkg/bucketeer/cache/processor/feature_flag_processor.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		return p.deleteAllAndSaveLocalCache(resp.FeatureFlagsId, resp.RequestedAt, resp.Features)
+	}
+	// Update only the updated flags
+	return p.updateLocalCache(
```
Comment (cre8ivejp) [inline pkg/bucketeer/cache/processor/feature_flag_processor.go:137]: When the `forceUpdate` is false, it only updates the updated and archived flags.

File: pkg/bucketeer/cache/processor/feature_flag_processor.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+		case <-p.closeCh:
+			return
+		}
+	}
+}
+
+func (p *processor) updateCache() error {
+	ftsID, err := p.getFeatureFlagsID()
+	if err != nil {
+		if err != cache.ErrNotFound {
```
Comment (Ubisoft-potato) [inline pkg/bucketeer/cache/processor/feature_flag_processor.go:109]: Comparison with errors using equality operators fails on wrapped errors, so it's better to use `!errors.Is(err, cache.ErrNotFound)`.

Comment (cre8ivejp) [reply pkg/bucketeer/cache/processor/feature_flag_processor.go]: Done.
https://github.com/bucketeer-io/go-server-sdk/pull/122/commits/3ff2e66d446694be1045e7223ef1a56d43cc84cd

**Review by Ubisoft-potato — APPROVED** [review]
LGTM! 👍

**Pull request merged**